### DISCUSSION TITLE
Is it possible to clone an output and create a new image using it?

### DISCUSSION THREAD

**Discussion by DOEHOONLEE** [ImageMagick/ImageMagick#4264] [Help] [answered]
Hi.

I was wondering if it is possible to clone an output and create a new image using it.

My steps are like this.

// convert specific layer in .psd to .png
convert 'original.psd[3]' 'convertedImg.png'

// then create an empty canvas with transparent background
xc:trasnparent -size 2000x2000

// clone the 'convertedImg.png' and place at 500 x 500
(-clone 0 -geometry +500+500) 'finalOutput.png'

This is how I tried, but getting `zsh: unknown file attribute:` error.
```bash
convert 'original.psd[2]' 'extracted.png' \
( +clone -size 2000x2000 -background blue -rotate 45) \
-geometry +500+500 -composite \
'Final.png'
```

or is there no way to clone an expected output image?

PS.

I tried exact codes that were provided as answers in [this post](https://legacy.imagemagick.org/discourse-server/viewtopic.php?t=32880), but getting the same `zsh: unknown file attribute:` error.

```bash
convert -size 4500x4500 xc:white ( -size 400x400 xc:red ) ^
( -clone 0 -clone 1 -geometry +500+900 -composite ) ^
( -clone 1 -background none -rotate 45 +repage ) ^
-delete 0,1 -geometry +500+900 -composite ^
a_new.png
```

Thank you!!

**Comment by snibgo** [comment]
What version of IM? If your use v7, I suggest you use "magick", not "convert".

Your shell seems to be zsh. The zsh syntax may be the same as for bash, but I don't know. You should consult a manual.

In bash, the command-continuation character (at the end of lines when the command continues) is backslash, "\\". Don't use caret "^", which is the Windows equivalent.

In bash, parentheses ( and ) are special characters that need to be escaped, like this: \\( and \\).

In ImageMagick, there must be a space on both sides of each (escaped) parenthesis.
```
magick 'original.psd[2]' 'extracted.png' \
\( +clone -size 2000x2000 -background blue -rotate 45 \) \
-geometry +500+500 -composite \
'Final.png'
```

**Reply by fmw42** [reply]
snibgo, I edited your post to make the backslashes show. You needed to escape the backslashes for them to show outside the <> syntax.

**Reply by snibgo** [reply]
Oos, sorry about that, and thanks for the correction.

**Reply by DOEHOONLEE** [reply]
![Screen Shot 2021-09-24 at 9 52 00 AM](https://user-images.githubusercontent.com/20305442/134602139-283ccfc8-636c-43dd-b047-e92c0a4b4e80.png)

I am using the stable release version 7. Thanks for pointing out to replace `convert` to `magick`.

This `-clone` is very interesting. I got the `-clone` to work now, but not in the way that I wanted to though.

Let me provide some images so that I can better explain.

**0. So this is the .psd and the layers.**
![photoshop](https://user-images.githubusercontent.com/20305442/134605133-10ce7a6f-0821-4183-b690-ea2dcb4fdb2c.png)
![layers](https://user-images.githubusercontent.com/20305442/134605148-afc4550d-dd7d-47d0-9ec2-f5752d37e029.png)

**1. When I run the command `magick 'sample.psd' 'sample.png'`, I get all individual layers converted to .png like this**
<img width="550" alt="Screen Shot 2021-09-24 at 10 38 34 AM" src="https://user-images.githubusercontent.com/20305442/134605371-e7591cb9-d57f-4e14-8ddf-bfa2174d1447.png">

0 being what the .psd looks like
1 is the deepest/lowest layer
then 2,3,4 in order.

**2. So let's say I would like to extract specific layer of .psd : smile => smile.png**
I run the following command `magick 'sample.psd[3]' 'smile.png'` then I get this smiley image.
![smile](https://user-images.githubusercontent.com/20305442/134605478-9b002e05-7687-4876-b74a-7191f8e54581.png)

**3. but because `imagemagick` only extracts the image in the layer, it does not get me the smile.png in the specific position where it used to be like this.**
![Screen Shot 2021-09-24 at 10 40 34 AM](https://user-images.githubusercontent.com/20305442/134605623-71484bcb-e7ae-432e-b8be-76c3f0b0429e.png)

**4. So I was planning to create an empty canvas with the size of original image (which is 2000x2000 here) then position smile with `geometry +x+y` ** 


```bash
convert 'original.psd[2]' 'smile.png' \
( +clone -size 2000x2000 -background white \) \
-geometry +500+500 -composite \
'Final.png'
```

is this still the right approach? or do I need to find another way?

Thank you!

**Comment by fmw42** [comment, marked as answer]
Post your PSD file. 

With IM 7, use magick, not convert.

You should be able to get it placed automatically with something like. I read the 0 (flattened) layer and the layer you want 2, then clone the first layer and fill it with white. Then swap the modified clone with the original 0 layer and delete the last image which is the original zero layer. Then I simply use -flatten to access the virtual canvas and write it onto the white image of size of the flattened image.

`magick original.psd[0,2] \( -clone 0 -fill white -colorize 100 \) -swap 0,2 +delete -background white -flatten final.png`

Does that work? If not, post your PSD file.

**Comment by fmw42** [comment]
You could also try:

`magick original.psd[0,2] -coalesce -background white -flatten final.png`

**Reply by DOEHOONLEE** [reply]
![Screen Shot 2021-09-24 at 11 07 25 AM](https://user-images.githubusercontent.com/20305442/134607558-09156c78-ef9f-44fa-bd87-4c0317deac49.png)

github comment does not allow me to upload .psd, but your solution works great!!

Let me write explanation of the codes you provided so, perhaps, it can be helpful to others too.

`magick original.psd[0,2] \( -clone 0 -fill white -colorize 100 \) -swap 0,2 +delete -background white -flatten final.png`

* .psd[0,2] extracts the [0] and [2] indexed layers in the .psd file
* -clone 0 clones the [0] image
* -fille white -colorize 100 => fills the background white with the rgb channel 100
* -swap => swaps the positions of the images
* -flatten merges the layers we have

`-coalesce` in the second solution you provided though, in the ImageMagick website, it says
![Screen Shot 2021-09-24 at 11 12 44 AM](https://user-images.githubusercontent.com/20305442/134607960-47bfc487-db47-4e89-ae0d-67578657ff3f.png)

but it can also be used to select specific index in the sequence in the process?

Thank you again for your help!

**Comment by fmw42** [comment]
**github comment does not allow me to upload .psd, but your solution works great!!**

Just zip the file and you can then upload.

**but it can also be used to select specific index in the sequence in the process?**

No, -coalesce just fills out the layers to their full size if the full dimensions are stored and can be retrieved. I have had trouble doing that with PSD files. So that was why I did the first solution. But I thought it might be worth a try. 

It is also used to fill out gif animation frames.

**Reply by DOEHOONLEE** [reply]
Oh, I did not know that. Thanks for the trick!!

**Comment by fmw42** [comment]
**-fille white -colorize 100 => fills the background white with the rgb channel 100**

Sorry, but I do not understand `rgb channel 100`.  -colorize 100 means fully colorize to 100% with that color. Look that up in the Options page and Anthony's Examples.

**Reply by DOEHOONLEE** [reply]
Oh, so it is like `opacity`?

**Reply by fmw42** [reply]
No, it does not control transparency. It just partially mixes or fully fills the image with a color over the original image.

**Reply by DOEHOONLEE** [reply]
Hmm I don't get what it means to fully/partially fill because when I create images with the percentage increment, the color gets darker.
<img width="663" alt="Screen Shot 2021-10-01 at 8 36 40 PM" src="https://user-images.githubusercontent.com/20305442/135613677-94e55b0e-aa0f-4cd8-92cf-5a5dfe5c68d7.png">

Is it like, all images are made up with pixels and it is about how many pixels to fill with -colorize?

**Comment by snibgo** [comment]
"-colorize" mixes (blends) the original colour of each pixel with the "-fill" colour. The value you give, eg 100, is the percentage of the fill colour in the mixture. For example, 50 would give a colour that was half way between the original colour and the fill colour. When the percentage is 100, the new colour is entirely the fill colour.

So "-fill White -colorize 100" makes all the pixels white.
